pandas/core/indexes/datetimes.py

```diff
@@ -49,12 +44,17 @@ def _new_DatetimeIndex(cls, d):
     # so need to localize
     tz = d.pop('tz', None)
 
-    with warnings.catch_warnings():
-        # we ignore warnings from passing verify_integrity=False
-        # TODO: If we knew what was going in to **d, we might be able to
-        #  go through _simple_new instead
-        warnings.simplefilter("ignore")
-        result = cls.__new__(cls, verify_integrity=False, **d)
+    if "data" in d and not isinstance(d["data"], DatetimeIndex):
```

Comment:
Without this we get a couple of failing pickle-based tests.  tests.frame.test_block_internals.test_pickle and tests.series.test_timeseries.test_pickle

Reply:
Gotcha, I've been fighting those too errors too. I suspect they were working before because DatetimeIndex accepts `data=None` for range-based constructors, which we don't want for the arrays.

Reply:
this is because you need to define ``__reduce__`` on in ``ExtensionArray``. This should *never* be hit by a non-DTI.

Reply:
I think you need to add back verfiy_integrity to the constructor

Reply:
I think that cases where validate_frequency raises are asking for trouble and the tests cases that fail this check are wonky.

@TomAugspurger you said you've battled with this.  Any thoughts?

Reply:
> and the tests cases that fail this check are wonky.

Yah, tracking one of these down, we are pickling+unpickling `df` defined:

```
@pytest.fixture
def timezone_frame()
    df = DataFrame({'A': date_range('20130101', periods=3),
                    'B': date_range('20130101', periods=3,
                                    tz='US/Eastern'),
                    'C': date_range('20130101', periods=3,
                                    tz='CET')})
    df.iloc[1, 1] = NaT
    df.iloc[1, 2] = NaT
    return df
```

The problem derives from the fact that we've called `__setitem__` (or an equivalent) and should be setting `freq=None` when we do so.

```
>>> dti = df['B']._values
>>> dti
DatetimeIndex(['2013-01-01 00:00:00-05:00', 'NaT',
               '2013-01-03 00:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq='D')

>>> DatetimeIndex._validate_frequency(dti, freq=dti.freq)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/pandas/core/indexes/datetimelike.py", line 230, in _validate_frequency
    raise ValueError(msg.format(infer=inferred, passed=freq.freqstr))
ValueError: Inferred frequency None from passed values does not conform to passed frequency D
```

Reply:
> @TomAugspurger you said you've battled with this. Any thoughts?
 
Medium-term, we can change `DatetimeArray.__new__` to `__init__` and all these problems go away (I think). We *may* not be able to do that until after the split though... I haven't tried.

Reply:
I assume Index subclasses are forced to use `__new__` since `Index.__new__` may return subclasses. But our arrays don't have that restriction.

Reply:
This is fixed by #24096

Reply:
let's fix that one first then. this needs to be changed here.

Reply:
The changes over in #24096 seem to be... different? I don't know how to explain it, but doesn't the fact that we're having to copy data over in #24096 seem disconnected from pickling?

Reply:
> but doesn't the fact that we're having to copy data over in #24096 seem disconnected from pickling?

Pickling turns out to be only tangentially related to the "real" problem.  In the status quo, altering a datetime64tz column alters the DatetimeIndex that backs it, but doesn't set its `freq` to `None`.  When that DataFrame is pickled and then unpickled, it tries to reconstruct that `DatetimeIndex`, but is passing arguments that _should_ raise `ValueError`.  ATM that gets side-stepped by passing `verify_integrity=False`.

So the goal of #24096 is to not corrupt the DatetimeIndex in the first place, making `verify_integrity=False` unnecessary.

That's still pretty roundabout.  Any clearer?

Reply:
Yeah, that does help.